### ISSUE TITLE
[dart:async] inconsistent behavior when synchronous future rethrows error

### ISSUE THREAD

**Issue by beeth0ven** [dart-lang/sdk#50050]
Hi there!

I have encountered an unexpected behavior with `Completer.sync()`, not sure if it's a bug or by design, let us start with minimal example:

1. using `Completer.sync` in general
```dart
void main()  {
  final completer = Completer<void>.sync();

  completer.future
    .then(
      (_) => print('onData'),
      onError: (_) => print('onError')
    );
  
  print('before completeError');
  completer.completeError(Exception());
  print('after completeError');
}
```
prints:
```
before completeError
onError
after completeError
```

this is exepected since it's synchronous.

2. rethrow error using `.catchError((..) => Futre.error(...))`

```diff
  final completer = Completer<void>.sync();
  
+ final autoCleanFuture = completer.future
+   .catchError((error, stackTrace) {
+     print('clean resources');
+     return Future.error(error, stackTrace);
+   });
    
- completer.future
+ autoCleanFuture
    .then(
      (_) => print('onData'),
      onError: (_) => print('onError')
    );
  
  print('before completeError');
  completer.completeError(Exception());
  print('after completeError');
```
prints:
```
before completeError
clean resources
after completeError
onError
```
as we can see `onError` been printed below `after completeError`, I expect it is above it.

3. rethrow errorr using `async/await`:

```diff
  final completer = Completer<void>.sync();
  
- final autoCleanFuture = completer.future
-   .catchError((error, stackTrace) {
-     print('clean resources');
-     return Future.error(error, stackTrace);
-   });

+ final autoCleanFuture = () async {
+   try {
+     await completer.future;
+   } catch (error) {
+     print('clean resources');
+     rethrow;
+   }
+ }();
    
  autoCleanFuture
    .then(
      (_) => print('onData'),
      onError: (_) => print('onError')
    );
  
  print('before completeError');
  completer.completeError(Exception());
  print('after completeError');
```
prints:
```
before completeError
clean resources
onError
after completeError
```

this is expected behavior since the `onError` is above `after completeError`.

So, I think case 1 and 3 is expected. I'm not sure if case 2 is a bug or i'm using it wrong. I care about it because this made my unit test fail while I expect it won't fail.

Any guide would be appreciated, thanks!

**Comment by mraleph** [comment]
The documentation for `Future.error` says "The created future will be completed with an error in a future microtask.". That's exactly what you see. 

I honestly would have expected 3 behave like 2. @lrhn is this expected that `Future` produced by `async` function completes eagerly like this?

**Comment by lrhn** [comment]
Everything here is within specification bounds, both language semantics and `Future` class documentation.

Number two is to be expected. You *create a new future*, which is initialized with an error, and which then completes asynchronously at a later point. It can't complete immediately where you create it, because the future has no listeners yet, so the error would immediately be an unhandled async error.
Printing `onError` has to wait for that extra asynchronous future.


Number three is an artifact of how `async` methods are implemented by the compilers/runtimes. The future returned by an `async` function must not complete with an error immediately, during the initial call to the `async` function and before the future has even been returned. Even if the body is entirely synchronous, there must be a delay, because otherwise (again) a future would be completed with an error before there has been any chance to add listeners to it.
So the implementation of `async` functions treats synchronous completion specially, adding an extra delay in the future completion step. Anything which happens after the first `await` is considered asynchronous, and it's safe to complete the returned future immediately. The `await`'s completion always happens as an asynchronous event (supposedly), so completing the returned future and propagating the result to its listeners can happen at the same time and still look like an asynchronous event too, one immediately following the previous one. And it's faster, so that's what the compiler does.

In this case, the completion of the `await` happened by calling a synchronous completer, not as a real event.
Any time you use an synchronous completer, you must *very carefully* consider whether the surrounding code is ready for arbitrary asynchronous events to happen *during* its execution, because they can.
Only use a synchronous completer if you know what you're doing, and make sure it only happens during the execution of synchronous code which is OK with being interrupted by asynchronous callbacks. Otherwise things can go spectacularly wrong. (Well, mostly your program crashing with an uncaught error, but sometimes you get subtle errors because code makes reasonable assumptions about interleaving that ends up not holding up, and it's very hard to debug.)

In short: Don't use a synchronous completer. (For experts only: Only use a synchronous completer where the code *cannot* tell that you did so.)

**Comment by beeth0ven** [comment]
I may need some time to absorb this valuable knownledge.
Thank you for clarification, mraleph and lrhn!
Thank you!
